### PR TITLE
[BEAM-22] Allow InProcess Evaluators to check Side Input completion

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This checks to ensure that the PCollectionView in the SideInputWindow
for the provided window either has elements available or is empty.

Schedule a future to ensure that the SideInputWindows are appropriately
filled with an empty iterable after retreiving the element.

This is allows the ParDoEvaluator to not attempt to process elements
that cannot currently be completed.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles @peihe

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java [hunk 1]
```diff
@@ -173,41 +176,43 @@ private void updatePCollectionViewWindowValues(
...
     private SideInputContainerSideInputReader(Collection<PCollectionView<?>> readerViews) {
       this.readerViews = ImmutableSet.copyOf(readerViews);
     }
 
     @Override
+    public boolean allViewsReadyInWindow(final BoundedWindow elementWindow) {
+      for (PCollectionView<?> view : readerViews) {
+        try {
+          BoundedWindow viewWindow =
+              view.getWindowingStrategyInternal().getWindowFn().getSideInputWindow(elementWindow);
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: I feel the conversion from the main input window to the side input window doesn't belong to the side input reader.
It is not consistent with get(...) method.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java [hunk 2]
```diff
@@ -320,18 +320,31 @@ public String getStepName(AppliedPTransform<?, ?, ?> application) {
...
 
   /**
+   * A {@link SideInputReader} that allows callers to check to see if the {@link PCollectionView}
+   * has contents in the specified window.
+   */
+  static interface ReadyCheckingSideInputReader extends SideInputReader {
+    /**
+     * Returns true if the {@link PCollectionView} is ready in the provided {@link BoundedWindow}.
+     */
+    boolean allViewsReadyInWindow(BoundedWindow window);
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: isAllViewsReadyInWindow(BoundedWindow sideInputWindow)?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: `isReady(view, sideInputWindow)` is symmetric with `get(view, sideInputWindow)`. Changed to that.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: changed the signature

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java [hunk 3]
```diff
@@ -173,41 +176,45 @@ private void updatePCollectionViewWindowValues(
...
+      checkArgument(
+          readerViews.contains(view),
+          "Tried to check if view %s was ready in a SideInputReader that does not contain it. "
+              + "Contained views; %s",
+          view,
+          readerViews);
+      try {
+        Future<Iterable<? extends WindowedValue<?>>> viewContents =
+            getViewFuture(view, window);
+        if (!viewContents.isDone()) {
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: return viewContents.isDone();
and remove return true at the very end

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java [hunk 4]
```diff
@@ -220,6 +227,38 @@ public String toString() {
       }
     }
 
+    /**
+     * Gets the future containing the contents of the provided {@link PCollectionView} in the
+     * provided {@link BoundedWindow}, setting up a callback to populate the future with empty
+     * contents if necessary.
+     */
+    private <T> Future<Iterable<? extends WindowedValue<?>>> getViewFuture(
+        final PCollectionView<T> view, final BoundedWindow window) throws ExecutionException {
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: If you wrap ExecutionException into a RuntimeException here, you can avoid to do it twice outside.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java [hunk 5]
```diff
@@ -320,18 +320,31 @@ public String getStepName(AppliedPTransform<?, ?, ?> application) {
   }
 
   /**
-   * Returns a {@link SideInputReader} capable of reading the provided
-   * {@link PCollectionView PCollectionViews}.
+   * Returns a {@link SideInputReader} capable of reading the provided {@link PCollectionView
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: update the javadoc for ReadyCheckingSideInputReader?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java [hunk 6]
```diff
@@ -320,18 +320,31 @@ public String getStepName(AppliedPTransform<?, ?, ?> application) {
...
    */
-  public SideInputReader createSideInputReader(final List<PCollectionView<?>> sideInputs) {
+  public ReadyCheckingSideInputReader createSideInputReader(
+      final List<PCollectionView<?>> sideInputs) {
     return sideInputContainer.createReaderForViews(sideInputs);
   }
 
   /**
+   * A {@link SideInputReader} that allows callers to check to see if the {@link PCollectionView}
+   * has contents in the specified window.
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: the javadoc sounds to me as describing isEmpty()

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 7]
```diff
@@ -327,16 +376,70 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+    ReadyCheckingSideInputReader reader =
+        container.createReaderForViews(ImmutableList.<PCollectionView<?>>of());
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("does not contain");
+    thrown.expectMessage(ImmutableList.of().toString());
+    reader.isReady(mapView, GlobalWindow.INSTANCE);
+  }
+
+  @Test
+  public void allViewsReadyInWindowForSomeNotReadyViewsFalseUntilElements() {
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: update the method name to isReady()

Also, do we have a convention to have a test prefix in the methods names, such as testIsReadyForSomeNotReadyViewsFalseUntilElements

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 8]
```diff
@@ -327,16 +376,70 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+                PaneInfo.ON_TIME_AND_ONLY_FIRING)));
+    assertThat(reader.isReady(mapView, SECOND_WINDOW), is(true));
+    assertThat(reader.isReady(singletonView, SECOND_WINDOW), is(true));
+
+    assertThat(reader.isReady(mapView, GlobalWindow.INSTANCE), is(false));
+    assertThat(reader.isReady(singletonView, GlobalWindow.INSTANCE), is(false));
+  }
+
+  @Test
+  public void allViewsReadyInWindowForEmptyWindowTrue() {
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: ditto

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 9]
```diff
@@ -327,16 +376,70 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+  }
+
+  @Test
+  public void allViewsReadyInWindowForSomeNotReadyViewsFalseUntilElements() {
+    container.write(
+        mapView,
+        ImmutableList.of(
+            WindowedValue.of(
+                KV.of("one", 1),
+                FIRST_WINDOW.maxTimestamp().minus(100L),
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: SECOND_WINDOW?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 10]
```diff
@@ -327,16 +376,70 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+        container.createReaderForViews(ImmutableList.of(mapView, singletonView));
+    assertThat(reader.isReady(mapView, SECOND_WINDOW), is(true));
+    assertThat(reader.isReady(singletonView, SECOND_WINDOW), is(false));
+
+    container.write(
+        singletonView,
+        ImmutableList.of(
+            WindowedValue.of(
+                1.25,
+                FIRST_WINDOW.maxTimestamp().minus(100L),
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: SECOND_WINDOW?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 11]
```diff
@@ -327,16 +376,70 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+        ImmutableList.of(
+            WindowedValue.of(
+                KV.of("one", 1),
+                FIRST_WINDOW.maxTimestamp().minus(100L),
+                SECOND_WINDOW,
+                PaneInfo.ON_TIME_AND_ONLY_FIRING)));
+
+    ReadyCheckingSideInputReader reader =
+        container.createReaderForViews(ImmutableList.of(mapView, singletonView));
+    assertThat(reader.isReady(mapView, SECOND_WINDOW), is(true));
```
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: add assert for the FIRST_WINDOW
write to the FIRST_WINDOW, and assert again?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: Using `getUnchecked` - our CacheLoader will never throw an exception.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java (hunk 6 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: Updated.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 7 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

not really.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 8 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 9 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 10 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 11 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java [hunk 12]
```diff
@@ -173,41 +176,32 @@ private void updatePCollectionViewWindowValues(
...
 
     @Override
+    public boolean isReady(final PCollectionView<?> view, final BoundedWindow window) {
+      checkArgument(
+          readerViews.contains(view),
+          "Tried to check if view %s was ready in a SideInputReader that does not contain it. "
+              + "Contained views; %s",
+          view,
+          readerViews);
+      Future<Iterable<? extends WindowedValue<?>>> viewContents = getViewFuture(view, window);
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: inline?
return getViewFuture(view, window).isDone();

**Comment by peihe** [comment]
LGTM

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java (hunk 12 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: Done.

**Comment by tgroh** [comment]
@kennknowles for committer

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 13]
```diff
@@ -327,16 +376,82 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
+    ReadyCheckingSideInputReader reader =
+        container.createReaderForViews(ImmutableList.<PCollectionView<?>>of());
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("does not contain");
+    thrown.expectMessage(ImmutableList.of().toString());
+    reader.isReady(mapView, GlobalWindow.INSTANCE);
+  }
+
+  @Test
+  public void isReadyForSomeNotReadyViewsFalseUntilElements() {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: javadoc this a little bit so one has context when parsing the test

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java [hunk 14]
```diff
@@ -327,16 +376,82 @@ public void finishDoesNotOverwriteWrittenElements() throws Exception {
...
             container.createReaderForViews(ImmutableList.<PCollectionView<?>>of(mapView)),
             mapView,
-            secondWindow);
+            SECOND_WINDOW);
 
     assertThat(mapFuture.get().isEmpty(), is(true));
   }
 
+  @Test
+  public void isReadyInEmptyReaderThrows() {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: The operative bit being that the window is not found, right? Not just that it is empty.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java [hunk 15]
```diff
@@ -220,6 +213,38 @@ public String toString() {
...
+            @Override
+            public void run() {
+              // The requested window has closed without producing elements, so reflect that in
+              // the PCollectionView. If set has already been called, will do nothing.
+              future.set(Collections.<WindowedValue<?>>emptyList());
+        }
+
+        @Override
+        public String toString() {
+          return MoreObjects.toStringHelper("InProcessSideInputContainerEmptyCallback")
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: `MoreObjects.toStringHelper(this)`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java [hunk 16]
```diff
@@ -320,18 +320,31 @@ public String getStepName(AppliedPTransform<?, ?, ?> application) {
...
+  public ReadyCheckingSideInputReader createSideInputReader(
+      final List<PCollectionView<?>> sideInputs) {
     return sideInputContainer.createReaderForViews(sideInputs);
   }
 
   /**
+   * A {@link SideInputReader} that allows callers to check to see if a {@link PCollectionView} has
+   * had its contents set in a window.
+   */
+  static interface ReadyCheckingSideInputReader extends SideInputReader {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: This interface seems somewhat generally useful, particularly for the [recent suggestion](http://mail-archives.apache.org/mod_mbox/incubator-beam-dev/201604.mbox/%3CCANMXwW3n6j9AJ9NT4xGejdi%3DuSDOhvG%2BU3xeGv0fpHiK%2ByhhSQ%40mail.gmail.com%3E) by @aljoscha for a shared streaming side input infrastructure. Have you and Pei already discussed maybe putting it top level alongside `SideInputReader`?

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessEvaluationContext.java]: This seems reasonable; potentially it's reasonable to add to SideInputReader directly (and then we can provide part of the shared runner code a factory that takes some sort of buffering strategy and uses it for non-ready elements)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java (hunk 15 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainer.java]: That'll just be some anonymous Runnable. I want the toString to be helpful in figuring out what it's doing.

Done, but via extracting to an inner class.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 13 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java (hunk 14 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/inprocess/InProcessSideInputContainerTest.java]: The Reader is empty - it does not contain the PCollectionView it is asked for, and thus should fail due to user error.

**Comment by kennknowles** [comment]
LGTM, merging.

**Pull request closed without merging**